topic_tools/src/tool_base_node.cpp

```diff
@@ -31,7 +31,7 @@ void ToolBaseNode::make_subscribe_unsubscribe_decisions()
 {
   if (auto source_info = try_discover_source()) {
     // always relay same topic type and QoS profile as the first available source
-    if (*topic_type_ != source_info->first || *qos_profile_ != source_info->second || !pub_) {
+    if (!topic_type_ || !qos_profile_ || *topic_type_ != source_info->first || *qos_profile_ != source_info->second || !pub_) {
```

Comment:
Or `||` operators in if statement doesn't guarantee order of operations for processor. 
Processor can execute `*topic_type_ != source_info->first || *qos_profile_ != source_info->second || !pub_` and than check `!topic_type_ || !qos_profile_` It seems fix doesn't fully address the problem.

Reply:
Thanks for your reply, @MichaelOrlov. Aren't logical OR as well as AND operators exceptions to the rule of unspecified order of evaluation in an expression? According to [cppreference.com](https://en.cppreference.com/w/cpp/language/eval_order): _Every value computation and side effect of the first (left) argument of the built-in [logical](https://en.cppreference.com/w/cpp/language/operator_logical) AND operator && and the built-in logical OR operator || is sequenced before every value computation and side effect of the second (right) argument._ Therefore, should any of our optionals here be uninitialized, the || operator will short-circuit and won't evaluate its following arguments.

Reply:
@bartek-kc Never mind, it seems after C++ 11 things got better and sequence of the or operations for || respecting as well.